File: Embassat/View/Artists/CADEMArtistsViewController.m
//
//  CADEMArtistsViewController.m
//  Embassat
//
//  Created by Joan Romano on 08/03/14.
//  Copyright (c) 2014 Crows And Dogs. All rights reserved.
//

#import "CADEMArtistsViewController.h"

#import "CADArrayDataSource.h"
#import "CADEMMenuCollectionViewCell.h"
#import "CADEMArtistDetailViewController.h"

#import "CADEMArtistsViewModel.h"
#import "CADEMArtistDetailViewModel.h"

@interface CADEMArtistsViewController ()

@property (nonatomic, weak) IBOutlet UICollectionView *artistsCollectionView;
@property (nonatomic, weak) IBOutlet UIActivityIndicatorView *activityIndicator;
@property (nonatomic, strong) CADArrayDataSource *dataSource;

@end

@implementation CADEMArtistsViewController

- (instancetype)init
{
    if (self = [super init])
    {
        self.title = @"ARTISTES";
    }
    
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.artistsCollectionView.dataSource = self.dataSource;
    [self.artistsCollectionView registerNib:[UINib nibWithNibName:@"CADEMMenuCollectionViewCell" bundle:nil] forCellWithReuseIdentifier:CADCellIdentifier];

    @weakify(self);
    [self.viewModel.updatedContentSignal subscribeNext:^(id x) {
        @strongify(self);
        [self.activityIndicator stopAnimating];
        [self.artistsCollectionView reloadData];
    }];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    self.viewModel.active = YES;
}

#pragma mark - UICollectionView Delegate

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath
{
    CADEMArtistDetailViewController *artistDetailViewController = [[CADEMArtistDetailViewController alloc] init];
    artistDetailViewController.viewModel = [self.viewModel artistViewModelForIndexPath:indexPath];
    
    [self.navigationController pushViewController:artistDetailViewController animated:YES];
}

#pragma mark - Lazy

- (CADArrayDataSource *)dataSource
{
    if (!_dataSource)
    {
        _dataSource = [[CADArrayDataSource alloc] initWithViewModel:self.viewModel configureCellBlock:^(CADEMMenuCollectionViewCell *cell, NSIndexPath *indexPath) {
            cell.optionName = [self.viewModel titleAtIndexPath:indexPath];
            cell.hidesTopSeparator = indexPath.row == 0;
        }];
    }
    
    return _dataSource;
}

@end

File: Embassat/View/Artists/CADEMArtistsViewController.swift
//
//  CADEMArtistsViewController.swift
//  Embassa't
//
//  Created by Joan Romano on 16/05/15.
//  Copyright (c) 2015 Crows And Dogs. All rights reserved.
//

import UIKit

public class CADEMArtistsViewController: CADEMRootViewController {
    
    @IBOutlet weak var artistsCollectionView: UICollectionView?
    @IBOutlet weak var activityIndicator: UIActivityIndicatorView?
    let dataSource: CADArrayDataSource
    let viewModel: CADEMArtistsViewModel
    
    override init(nibName nibNameOrNil: String?, bundle nibBundle: NSBundle?) {

        let theViewModel = CADEMArtistsViewModel()
        dataSource = CADArrayDataSource(viewModel: theViewModel, configureCellBlock: { (cell: AnyObject!, indexPath: NSIndexPath) -> Void in
            let theCell = cell as! CADEMArtistCollectionViewCell
            
            theCell.optionName = theViewModel.titleAtIndexPath(indexPath)
            theCell.hidesBottomSeparator = indexPath.row == theViewModel.numberOfItemsInSection(0) - 1
        }, configureHeaderBlock: nil)
        viewModel = theViewModel
        
        super.init(nibName: nibNameOrNil, bundle: nibBundle)
    }

    required public convenience init(coder aDecoder: NSCoder) {
        self.init(nibName: nil, bundle: nil)
    }

    override public func viewDidLoad() {
        super.viewDidLoad()

        title = "Artistes"
        artistsCollectionView?.dataSource = self.dataSource
        artistsCollectionView?.registerNib(UINib(nibName: "CADEMArtistCollectionViewCell", bundle: nil), forCellWithReuseIdentifier: CADArrayDataSource.CADCellIdentifier)
        
        if viewModel.numberOfItemsInSection(0) > 0 {
            self.activityIndicator?.stopAnimating()
        }
        
        viewModel.activeSubject.subscribeNext({ [unowned self] (_) -> Void in
            self.activityIndicator?.stopAnimating()
            self.artistsCollectionView?.reloadData()
        }, error: { (_) -> Void in
            self.activityIndicator?.stopAnimating()
            self.artistsCollectionView?.reloadData()
        })
    }
    
    public override func viewDidDisappear(animated: Bool) {
        super.viewDidDisappear(animated)
        
        if let indexPaths = artistsCollectionView?.indexPathsForSelectedItems() {
            for indexPath in indexPaths {
                artistsCollectionView?.deselectItemAtIndexPath(indexPath, animated: false)
            }
        }
    }

    func collectionView(collectionView: UICollectionView, didSelectItemAtIndexPath indexPath: NSIndexPath) {
        let artistViewController = CADEMArtistDetailViewController(nibName: "CADEMArtistDetailViewController", bundle: nil)
        artistViewController.viewModel = viewModel.artistViewModel(forIndexPath: indexPath)
        artistViewController.updateSignal.subscribeNext { [unowned self] (_) -> Void in
            self.viewModel.shouldRefreshModel()
        }
        
        self.navigationController?.pushViewController(artistViewController, animated: true)
    }
}
